### PR TITLE
Added module for basic feature selection capabilities and updated set…

### PR DESCRIPTION
**Pull request by pediejo**
…up file to require cvxopt

### CODE REVIEW

File: src/python/zquantum/featureselection/feature_selection.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
+import numpy as np
+from cvxopt import matrix, solvers
+from scipy.stats import pearsonr
+from scipy.optimize import minimize, LinearConstraint
+from sklearn.feature_selection import mutual_info_classif, mutual_info_regression
+
+def construct_pearson_corr_relevance_vector(feature_matrix, label_vector):
```
[the remark this answers is not in the document]
Comment (pediejo) [reply src/python/zquantum/featureselection/feature_selection.py]: https://github.com/zapatacomputing/z-quantum-feature-selection/pull/2/commits/a0d697e3224900f6f6652b6d49b41ee3ecda410e

and 

https://github.com/zapatacomputing/z-quantum-feature-selection/pull/2/commits/3c2441c86e5d8e5d6f24858e07e0427e42e8bf9b

File: src/python/zquantum/featureselection/feature_selection.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    """Constructs the Pearson correlation relevance vector of a feature matrix and vector of discrete labels.
+        This follows the construction of F on pg. 1495 of 'Rodriguez-Lujan, Irene, et al. "Quadratic programming 
+        feature selection." Journal of Machine Learning Research (2010).'
+
+    Args:
+        feature_matrix (np.ndarray): 2D array with features along the columns and data points along the rows.
+        label_vector (np.ndarray of ints): 1D array with each data point assigned to a column.
```
[the remark this answers is not in the document]
Comment (pediejo) [reply src/python/zquantum/featureselection/feature_selection.py]: https://github.com/zapatacomputing/z-quantum-feature-selection/pull/2/commits/609e3af506f527f64dcd6348c46b97d8748fcb57

File: src/python/zquantum/featureselection/feature_selection.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+    if seed is not None:
+        np.random.seed(seed=seed)
+
+    relevancy_vector =  mutual_info_classif(feature_matrix, label_vector)
+
+    return relevancy_vector
+
+
+def construct_pearson_corr_redundancy_matrix(feature_matrix):
```
[the remark this answers is not in the document]
Comment (pediejo) [reply src/python/zquantum/featureselection/feature_selection.py]: Added typing
https://github.com/zapatacomputing/z-quantum-feature-selection/commit/a0d697e3224900f6f6652b6d49b41ee3ecda410e

File: src/python/zquantum/featureselection/feature_selection.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        redundancy_matrix (np.ndarray): 2D square array assigning a correlation score to each pair of features.
+        relevancy_vector (np.ndarray): 1D array assigning a relevancy score to each feature.
+        alpha (float): parameter between 0 and 1 which weights the importance of relevancy (towards 0) vs redundancy (towards 1).
+
+    Returns:
+        weight_vector (np.ndarray): 1D array assigning an importance weight to each feature.
+    """     
+
+    num_of_features = len(relevancy_vector)
+    P = matrix(redundancy_matrix*(1.0-alpha))
```
[the remark this answers is not in the document]
Comment (pediejo) [reply src/python/zquantum/featureselection/feature_selection.py]: https://github.com/zapatacomputing/z-quantum-feature-selection/pull/2/commits/256b08dff0afb828239aac74f8694673f3d1e51e

File: src/python/zquantum/featureselection/feature_selection.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+    # Set alpha as default if input is None
+    if alpha is None:
+        bar_q = np.mean(redundancy_matrix)
+        bar_f = np.mean(relevancy_vector)
+        alpha = bar_q / (bar_q + bar_f)
+
+    feature_weights = _weight_features_with_quadratic_programming(redundancy_matrix, relevancy_vector, alpha)
+    ranking = np.argsort(feature_weights)[::-1]
+    chosen_ones = set(ranking[:num_of_chosen_features])
```
[the remark this answers is not in the document]
Comment (pediejo) [reply src/python/zquantum/featureselection/feature_selection.py]: https://github.com/zapatacomputing/z-quantum-feature-selection/pull/2/commits/1bd84e11e17d70e691c069a318172e89b1fa5b17

File: src/python/zquantum/featureselection/feature_selection_test.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+from feature_selection import (construct_mutual_info_relevance_vector,
+                               construct_pearson_corr_relevance_vector,
+                               construct_mutual_info_redundancy_matrix,
+                               construct_pearson_corr_redundancy_matrix,
+                               quadratic_programming_feature_selection,
+                               greedy_mrmr_feature_selection)
+
+# The dataset consists of 10 points around (0, 0, 0) and 10 points around (10, 10, 10)
+# The feature vector for point (x, y, z) is (x, y, z, sin(x))
+X = np.array([[-0.8,  -1.04,  0.53, -0.72],
```
[the remark this answers is not in the document]
Comment (pediejo) [reply src/python/zquantum/featureselection/feature_selection_test.py]: https://github.com/zapatacomputing/z-quantum-feature-selection/pull/2/commits/5315a87d3cb0ae164627ffcdfb58bb91e91b1b67

**Review by mstechly — APPROVED** [review]

**Pull request merged**